coordinator/core/marbleapi.go

```diff
@@ -112,6 +114,20 @@ func (c *Core) Activate(ctx context.Context, req *rpc.ActivationReq) (*rpc.Activ
 	if marble.Parameters == nil {
 		marble.Parameters = &rpc.Parameters{}
 	}
+
+	// Union user-defined secrets with shared and unique secrets
+	userSecrets := findUserSecrets(marble.Parameters, mainManifest.Secrets)
+	for _, k := range userSecrets {
+		secret, err := c.data.getSecret(k)
+		if err != nil {
+			if store.IsStoreValueUnsetError(err) {
+				return nil, fmt.Errorf("secret %s needed but not set", k)
+			}
+			return nil, err
+		}
+		secrets[k] = secret
+	}
```

Comment:
```suggestion
	if err := c.addUserSecrets(secrets, marble.Parameters, mainManifest.Secrets); err != nil {
		return err
	}
```
we should keep the Activate func rather high level, delegating tasks to subfunctions